src/plugins/intel_gpu/src/graph/impls/ocl/kernel_selector_helper.h

```diff
@@ -300,7 +300,7 @@ inline void update_shapes(kernel_selector::Params& p, const kernel_impl_params&
         auto& fd = bp.fused_ops[i];
         fd.output_tensor = convert_data_tensor(fused_prim.output_layout);
         for (size_t i = fd.dep_idx_start; i < fd.dep_idx_start + fd.dep_size; i++) {
-            fd.tensors.push_back(convert_data_tensor(impl_param.get_input_layout(i)));
+            fd.AddTensor(convert_data_tensor(impl_param.get_input_layout(i)));
```

Comment:
I think data race is unexpected here. kernel_impl_params object should be unique for each primitive in each thread. Could you describe a little bit what causes data race?

Reply:
Hi @vladimir-paramuzov, previously, I found that **update_shapes** was called **twice** from the debug log:
```
--> primitive_inst: multiply:Multiply_22381, execute update_impl start
--> primitive_inst: multiply:Multiply_22381, execute update_impl start
[+] update_shapes
[+] update_shapes
```
and it will fail at **fd.tensors.push_back**, since 1 of them updated the fd.tensors data pointer:
(so the second "[after   ] ***" doesn't printed out in the log)
```
[before] fd.tensors.size: 0x10, fd.tensors.data: 0x7ffd2484f960
[before] fd.tensors.size: 0x10, fd.tensors.data: 0x7ffd2484f960
[after ] fd.tensors.size: 0x11, fd.tensors.data: 0x7ffd24a88840
```

I think this may be caused by my 2 Arc A770 cards setting on the machine.
After changing to only 1 iGPU on another machine, I can't re-produce the **twice** called **update_shapes** in my debug log now, so update_shapes is only called once (for primitive) and should not be the problem. But the segmentation fault issue still existed.

After excluding update_shapes, I will try to track the fd to see if it is used simultaneously somewhere else. (to see the potential cause of data race). 
BTW, my commit in this PR also works for the iGPU (as well as previous dGPU A770 when debugging).

Reply:
_update_shapes is only called once (for primitive) and should not be the problem. But the segmentation fault issue still existed._
Could you go deeper why it happens

Reply:
@p-durandin Yes, this issue happens when running multiple times "start_async()" in the test_cross-lingual-books-alignment.ipynb. notebook.
Changing multiple times running of "start_async()" to "infer()" will have no issue, so the problem is related to multiple asynchronized running.

Reply:
Added a log to detect if concurrent access happen.  Yes, it indeed happened when running in notebook.  benchmark_app doesn't reproduce this issue with many nstreams.  

    for (size_t i = 0; i < bp.fused_ops.size(); i++) {
        const auto& fused_prim = impl_param.fused_desc[i];
        auto& fd = bp.fused_ops[i];
        if (!vector_mutex.try_lock()) {
            log_message("WARNING: Concurrent access detected! Another thread is accessing the vector.");
        }
        fd.output_tensor = convert_data_tensor(fused_prim.output_layout);
        for (size_t i = fd.dep_idx_start; i < fd.dep_idx_start + fd.dep_size; i++) {
fd.tensors.push_back(convert_data_tensor(impl_param.get_input_layout(i)));
        }
        if (vector_mutex.try_lock()) { // Check if this thread holds the lock
            vector_mutex.unlock();
        }
    }

Another suggestion to the fix.  Instead using mutex, change the std::vector to concurrent_vector.

src/plugins/intel_gpu/src/kernel_selector/kernel_selector_params.h

#include <tbb/concurrent_vector.h>

//using MultiDataTensor = std::vector<DataTensor>;
using MultiDataTensor = tbb::concurrent_vector<DataTensor>;

Reply:
[2024-11-06 18:51:51.444] [Thread **135250088822336**] WARNING: Concurrent access detected! Another thread is accessing the vector.
[2024-11-06 18:51:51.449] [Thread **135245015811648**] WARNING: Concurrent access detected! Another thread is accessing the vector.
[2024-11-06 18:51:51.449] [Thread **135245015811648**] WARNING: Concurrent access detected! Another thread is accessing the vector.

Reply:
Thanks @clee30 & @sshlyapn for your suggestions.

Dumped the model properties for the model, and seems the numbers of **infer requests** & **streams** are different:
(left: with performance hint, failed; right: without performance hint, succeeded)
![image](https://github.com/user-attachments/assets/80156fb9-2060-4df0-b546-998ab0767c85)

Reply:
With `config={hints.performance_mode(): hints.PerformanceMode.THROUGHPUT}`, the NUM_STREAMS will be 2. 
```
PERFORMANCE_HINT: PerformanceMode.THROUGHPUT
EXECUTION_MODE_HINT: ExecutionMode.PERFORMANCE
COMPILATION_NUM_THREADS: 20
NUM_STREAMS: 2
PERFORMANCE_HINT_NUM_REQUESTS: 0
INFERENCE_PRECISION_HINT: <Type: 'float16'>
DYNAMIC_QUANTIZATION_GROUP_SIZE: 32
ACTIVATIONS_SCALE_FACTOR: 0.0
DEVICE_ID: 0
EXECUTION_DEVICES: ['GPU.0']
```
And 2 different threads (thread `729e28c00640` and `729e17e00640`) will be used to start_async:
```
--> sentence idx = 0 start_async
--> cpu_streams_executor Impl run task with thread: 729e28c00640
--> sentence idx = 1 start_async
--> cpu_streams_executor Impl run task with thread: 729e17e00640
```
These 2 threads will running concurrently and access the same fd (fused_op_description: `0x729bdc004490`) and push_back new element to the same fd.tensors (std::vector: `0x729c0c040850`).
```
--> update_shapes[729e28c00640] i: 0, &fd: 0x729bdc004490, id: mvn:__module.encoder.layer.0.attention.output.LayerNorm/aten::layer_norm/MVN, fd.tensors size: 3, ptr: 0x729c0c040850
--> update_shapes[729e17e00640] i: 0, &fd: 0x729bdc004490, id: mvn:__module.encoder.layer.0.attention.output.LayerNorm/aten::layer_norm/MVN, fd.tensors size: 4, ptr: 0x729c0c040850
```
And thus there is data race of fd.tensors under this THROUGHPUT mode.